src/login/logind.c

```diff
@@ -740,18 +739,17 @@ static int manager_connect_bus(Manager *m) {
         if (r < 0)
                 return log_error_errno(r, "Failed to request match for Reloading: %m");
 
-        r = sd_bus_call_method(
+        r = sd_bus_call_method_async(
                         m->bus,
+                        NULL,
                         "org.freedesktop.systemd1",
                         "/org/freedesktop/systemd1",
                         "org.freedesktop.systemd1.Manager",
                         "Subscribe",
-                        &error,
-                        NULL, NULL);
-        if (r < 0) {
-                log_error("Failed to enable subscription: %s", bus_error_message(&error, r));
-                return r;
-        }
```

Comment:
bus messages are strictly ordered. That means once our Subscribe call is queued it is guaranteed to be dispatched before our enumeration would start. It's one of the nicer properties of D-Bus actually: the order you submit your jobs is the order the messages are dispatched (at least as you only talk to a single server and the driver only)

Reply:
I meant the enumeration that happens based on files on disk. The ordering of dbus messages is asynchronous wrt. to any operations we do.

Reply:
not following, still? can you elaborate?

Reply:
Sorry for the delay. It's possible I'm being dense, but...:
in manager_startup() we call manager_connect_bus() first and later enumerate various seats and users by reading state from disk. manger_connect_bus() subscribes to various signals to monitor the addition and removal of users and seats. IIUC, if a user was found in the state saved on disk, and is then removed any time after we enabled the subscriptions, logind will be notified and will update it's internal state to drop the user. Previously, subscriptions happened before the enumeration. But now that the subscriptions are made asynchronous, they could happen after the enumeration, which opens a window for the user to be removed without logind being notified.

Reply:
manager_startup() enqueues Subscribe() to the bus, then enumerates stuff on disk, and then calls manager_gc() which goes through all open sessions and checks via the bus whether their scope units are still there, reaping them when not. Since the checking is done after the Subscribe() all is good. What matters is that we Subscribe first, and GC second. Whether the GC is async or not doesn't matter, given that it must be dispatched before the GC, due to dbus ordering constraints, and the GC check is sync.